### PR TITLE
task4: implemented task due to requirements

### PR DESCRIPTION
**Pull request by klimashevsky-pavel**
task: https://epam-my.sharepoint.com/personal/vitali_kozlov_epam_com/_layouts/15/onedrive.aspx?id=%2Fpersonal%2Fvitali_kozlov_epam_com%2FDocuments%2FNode%2Ejs%2FCDP%2FHomeworks%2F4%2E%20Middleware%2E%20Frameworks%2FHomework%204%2Epdf&parent=%2Fpersonal%2Fvitali_kozlov_epam_com%2FDocuments%2FNode%2Ejs%2FCDP%2FHomeworks%2F4%2E%20Middleware%2E%20Frameworks
Due to not fully correct task requirements, some things were changed (used usual middlewares for express, routes respond without any CRUD operations)

### CODE REVIEW

File: src/app.ts [hunk 1]
```diff
@@ -1,45 +1,59 @@
...
+import 'http-servers/plain-text-server';
+import 'http-servers/html-server';
+import 'http-servers/json-server';
+import 'http-servers/echo-server';
+
+// run file watchers from previous tasks
+runWatchers();
+
+// Made Server singleton mostly for learning purposes
+class Server {
```
Comment (deser) [inline src/app.ts:19]: Module is a singleton by itself. Class seems here burdensome

File: src/http-servers/json-server.ts [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+    } catch (e) {
+        logger.error(e);
+    }
+    res.writeHead(200, headerParams);
+    res.end(body);
+};
+
+const server = http.createServer(requestHandler);
+
+server.listen(PORT, (err: Error) => {
```
Comment (deser) [inline src/http-servers/json-server.ts:30]: Good practice it to handle shutdown gracefully

File: src/http-servers/html-server.ts [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+const requestHandler = (req: IncomingMessage, res: ServerResponse) => {
+    const htmlPath = path.join(__dirname, 'index.html');
+    const customReadableStream = createCustomReadableStream(htmlPath);
+    res.writeHead(200, headerParams);
+    customReadableStream.pipe(res);
+};
+
+const server = http.createServer(requestHandler);
+
+server.listen(PORT, (err: Error) => {
```
Comment (deser) [inline src/http-servers/html-server.ts:41]: Good practice it to handle shutdown gracefully

File: src/http-servers/echo-server.ts [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+const PORT = 3004;
+
+const requestHandler = (req: IncomingMessage, res: ServerResponse) => {
+    res.writeHead(200);
+    req.pipe(res);
+};
+
+const server = http.createServer(requestHandler);
+
+server.listen(PORT, (err: Error) => {
```
Comment (deser) [inline src/http-servers/echo-server.ts:12]: Good practice it to handle shutdown gracefully

File: src/http-servers/plain-text-server.ts [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+
+const requestHandler = (req: IncomingMessage, res: ServerResponse) => {
+    const body = 'Hello World';
+    res.writeHead(200, headerParams);
+    res.end(body);
+};
+
+const server = http.createServer(requestHandler);
+
+server.listen(PORT, (err: Error) => {
```
Comment (deser) [inline src/http-servers/plain-text-server.ts:16]: Good practice it to handle shutdown gracefully

File: src/index.ts [hunk 6]
```diff
@@ -0,0 +1,6 @@
+import app from 'app';
+
+const PORT = process.env.PORT || 8080;
+app.listen(PORT, () => {
```
Comment (deser) [inline src/index.ts:4]: Good practice it to handle shutdown gracefully

File: src/interfaces/ExtendedRequest.ts [hunk 7]
```diff
@@ -0,0 +1,9 @@
+import { UrlWithParsedQuery } from 'url';
```
Comment (deser) [inline src/interfaces/ExtendedRequest.ts:1]: implement yours

**Review by deser — CHANGES_REQUESTED** [review]

**Review by deser — CHANGES_REQUESTED** [review]
Implement cookie and query prasers

File: src/interfaces/ExtendedRequest.ts (hunk 7 above)
Comment (sleeping-cucumber) [reply src/interfaces/ExtendedRequest.ts:1]: this is Node native module, no problems with using it

File: src/app.ts (hunk 1 above)
Comment (sleeping-cucumber) [reply src/app.ts:19]: I like using OO style and I don't see any problems with class here. Maybe you are right about singleton here, but I'be decided to implement it for learning purposes.

**Pull request merged**